extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/FlywayProducer.java

```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright 2019 Red Hat, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.quarkus.flyway.runtime;
+
+import javax.enterprise.context.ApplicationScoped;
+import javax.enterprise.context.Dependent;
+import javax.enterprise.inject.Produces;
+
+import org.flywaydb.core.Flyway;
+import org.flywaydb.core.api.configuration.FluentConfiguration;
+
+import io.quarkus.agroal.runtime.AgroalRuntimeConfig;
+
+@ApplicationScoped
+public class FlywayProducer {
+    private AgroalRuntimeConfig agroalRuntimeConfig;
+    private FlywayRuntimeConfig flywayRuntimeConfig;
+    private FlywayBuildConfig flywayBuildConfig;
+
+    @Produces
+    @Dependent
+    public Flyway produceFlyway() {
+        FluentConfiguration configure = Flyway.configure();
+        String url = agroalRuntimeConfig.defaultDataSource.url
+                .orElseThrow(() -> new IllegalStateException(
+                        "No default datasource configured. Please specify the quarkus.datasource.* properties"));
+        String username = agroalRuntimeConfig.defaultDataSource.username.orElse(null);
+        String password = agroalRuntimeConfig.defaultDataSource.password.orElse(null);
+        configure.dataSource(url, username, password);
```

Comment:
Couldn't you inject the Agroal datasource directly? I see Flyway has a method for this: https://github.com/flyway/flyway/blob/master/flyway-core/src/main/java/org/flywaydb/core/api/configuration/FluentConfiguration.java#L735 .

Reply:
I tried but it didn't work so I moved on with the config alternative. 
I got an UnsatisfiedResolutionException when I tried to inject it.

```
Caused by: javax.enterprise.inject.UnsatisfiedResolutionException: Unsatisfied dependency for type io.agroal.api.AgroalDataSource and qualifiers [@Default]
        - java member: io.quarkus.flyway.runtime.FlywayProducer#dataSource
        - declared on CLASS bean [types=[io.quarkus.flyway.runtime.FlywayProducer, java.lang.Object], qualifiers=[@Default, @Any], target=io.quarkus.flyway.runtime.FlywayProducer]
        at io.quarkus.arc.processor.Beans.resolveInjectionPoint(Beans.java:326)
        at io.quarkus.arc.processor.BeanInfo.init(BeanInfo.java:375)
        at io.quarkus.arc.processor.BeanDeployment.init(BeanDeployment.java:268)

```
![image](https://user-images.githubusercontent.com/1218220/56472832-22b16a80-6429-11e9-9609-5e262744c5ef.png)

Reply:
This is weird.

@mkouba a producer should be able to use beans created by other producers or it's not possible?

Reply:
@gsmet Yes, it should be possible to inject a bean either directly in a producer method param or in the bean that declares the producer method.

Reply:
Is the default `AgroalDataSource` present?

Reply:
AFAICS here https://github.com/quarkusio/quarkus/pull/2095/files#diff-379368a12549672304fea09d1785f1feR20 , it should be. But I haven't tried to understand what's going on yet.

Reply:
@gsmet @cristhiank It seems that the injection fails in `FlywayExtensionFullConfigTest` because `quarkus-flyway-deployment` does not depend on `quarkus-agroal-deployment`, I believe it should.

Reply:
@cristhiank did you see @mkouba's suggestion? Can you check it solves the issue?

Reply:
@gsmet  Yes, I just tried at this very moment. But now I get this:

```
Caused by: java.lang.IllegalStateException: The datasources are not ready to be consumed: the runtime configuration has not been injected yet
        at io.quarkus.agroal.runtime.AbstractDataSourceProducer.checkRuntimeConfig(AbstractDataSourceProducer.java:195)
        at io.quarkus.agroal.runtime.AbstractDataSourceProducer.getDefaultRuntimeConfig(AbstractDataSourceProducer.java:63)
        at io.quarkus.agroal.runtime.DataSourceProducer.createDefaultDataSource(Unknown Source)
        at io.quarkus.agroal.runtime.DataSourceProducer_ProducerMethod_createDefaultDataSource_7c487e3ef869f878aa871e917c94f4d26d5d5c56_Bean.create(Unknown Source)
        at io.quarkus.agroal.runtime.DataSourceProducer_ProducerMethod_createDefaultDataSource_7c487e3ef869f878aa871e917c94f4d26d5d5c56_Bean.create(Unknown Source)
        at io.quarkus.arc.AbstractSharedContext.createInstanceHandle(AbstractSharedContext.java:86)
        at io.quarkus.arc.AbstractSharedContext.lambda$new$0(AbstractSharedContext.java:33)
        at io.quarkus.arc.ComputingCache$CacheFunction.lambda$apply$0(ComputingCache.java:115)

```

At least it tries to inject it. But something is still missing.

Reply:
OK, I'll take a look, we probably need to sync things somehow.

Reply:
I was able to solve it by using `Instance<AgroalDataSource>` instead of inject it as a field. I don't know if it is the canonical way to do it. I just commited the changes for your review. @gsmet @mkouba

Reply:
I am still using the Lazy `Instance<>` to inject the datasource because I wasn't able to inject the `AgroalDatasource` directly. 

After some debugging I believe it is related to the execution order for the `BeanContainerListener`s 

see:

https://github.com/quarkusio/quarkus/blob/a31c4264d962a008d0d776ea0a44ba8f6dab95f0/extensions/arc/runtime/src/main/java/io/quarkus/arc/runtime/ArcDeploymentTemplate.java#L107-L112

![image](https://user-images.githubusercontent.com/1218220/56843051-bcca4600-6861-11e9-9ece-ccf93081fff9.png)

The FlywayTemplate is being executed before the AgroalTemplate, so the AgroalConfig is not assigned when Flyway tries to use it. 😖

Reply:
This is fixed in master. I was able to inject the AgroalDatasource directly